java/ql/src/semmle/code/java/security/RequestForgery.qll

```diff
@@ -0,0 +1,334 @@
+ */
+private class SpringRestTemplateUrlMethod extends Method {
+  SpringRestTemplateUrlMethod() {
+    this.getDeclaringType() instanceof SpringRestTemplate and
+    this.hasName([
+        "doExecute", "postForEntity", "postForLocation", "postForObject", "put", "exchange",
+        "execute", "getForEntity", "getForObject", "patchForObject"
+      ])
+  }
+
+  /**
+   * Gets the argument which corresponds to a URL argument
+   * passed as a `java.net.URL` object or as a string or the like
+   */
+  Argument getUrlArgument(MethodAccess ma) {
+    // doExecute(URI url, HttpMethod method, RequestCallback requestCallback,
+    //  ResponseExtractor<T> responseExtractor)
+    result = ma.getArgument(0)
+  }
+}
+
+/** A sanitizer for request forgery vulnerabilities. */
+abstract class RequestForgerySanitizer extends DataFlow::Node { }
+
+private class PrimitiveSanitizer extends RequestForgerySanitizer {
+  PrimitiveSanitizer() { this.getType() instanceof PrimitiveType }
+}
+
+private class HostnameSanitizingPrefix extends CompileTimeConstantExpr {
+  int offset;
+
+  HostnameSanitizingPrefix() {
+    // Matches strings that look like when prepended to untrusted input, they will restrict
+    // the host or entity addressed: for example, anything containing `?` or `#`, or a slash that
+    // doesn't appear to be a protocol specifier (e.g. `http://` is not sanitizing), or specifically
+    // the string "/".
+    exists(
+      this.getStringValue()
+          .regexpFind(".*([?#]|[^?#:/\\\\][/\\\\]).*|[/\\\\][^/\\\\].*|^/$", 0, offset)
+    )
+  }
+
+  /**
+   * Gets the offset in this constant string where a sanitizing substring begins.
+   */
+  int getOffset() { result = offset }
+}
+
+private AddExpr getParentAdd(AddExpr e) { result = e.getParent() }
+
+private AddExpr getAnAddContainingHostnameSanitizingPrefix() {
+  result = getParentAdd*(any(HostnameSanitizingPrefix p).getParent())
+}
+
+private Expr getASanitizedAddOperand() {
+  exists(AddExpr e |
+    e = getAnAddContainingHostnameSanitizingPrefix() and
+    (
+      e.getLeftOperand() = getAnAddContainingHostnameSanitizingPrefix() or
+      e.getLeftOperand() instanceof HostnameSanitizingPrefix
+    ) and
+    result = e.getRightOperand()
+  )
+}
+
+private MethodAccess getNextAppend(MethodAccess append) {
+  result = any(StringBuilderVar sbv).getNextAppend(append)
+}
+
+private Expr getQualifier(MethodAccess e) { result = e.getQualifier() }
+
+private MethodAccess getAChainedAppend(Expr e) {
+  (
+    result.getQualifier() = e
+    or
+    result.getQualifier() = getAChainedAppend(e)
+  ) and
+  result.getCallee().getDeclaringType() instanceof StringBuildingType and
+  result.getCallee().getName() = "append"
+}
+
+/**
+ * An expression that is sanitized because it is concatenated onto a string that looks like
+ * a hostname or a URL separator, preventing the appended string from arbitrarily controlling
+ * the addressed server.
+ */
+private class HostnameSanitizedExpr extends Expr {
```

Comment:
It seems like a lot of these cases can be simplified significantly with the following insight: In a concatenation `A + B` where `A` is a safe prefix, there's no need to sanitise `B` - it is fully sufficient to sanitise the result of the addition, i.e. the entire `A + B` expression. This generalises to `StringBuilder.append` and `String.format` where we don't need to sanitise the tainted data being appended as long as we just sanitise the result.

Reply:
I don't think this insight does apply in any of those cases:

* For addition, you have the case `(userSupplied + "?") + otherUserSupplied`. The expression `userSupplied + "?"` is tainted, but also sanitises `otherUserSupplied` because it constitutes a sanitising prefix. Thus we sanitise the right-hand operand to the add, not the result, thereby preserving any taint already present on the LHS.
* For string-building, sanitising the result has no benefit as the expressions that would add taint (and which should be sanitised) are coming in the future -- hence the logic to walk forwards in chains of append operations.
* For String.format and cousins we have multiple inputs and one result, so we must sanitise the particular inputs that fall to the right of the sanitising prefix